### ISSUE TITLE
[abstracts] transitivity of implicit casts

### ISSUE THREAD

**Issue by vonagam** [HaxeFoundation/haxe#9694]
Example:
```haxe
abstract A(Int) from Int to Int {}

abstract B(A) from A to A {}

abstract C(B) from B to B {}

abstract X(Int) from Int {
  @:from static function fromInt(value: Int): X return value;
}

class Main {
  static public function main() {
    var ai: A = 1; // Int -> A works
    var ci: C = 2; // Int -> A -> B -> C works
    var ca: C = ai; // A -> B -> C fails: A should be C
    var xc: X = ci; // C -> B -> A -> Int -> X works
  }
}
```

While working on #9691, learned that there are some cases where abstract implicit casts are transitive. 

Are they supposed to be in those cases (`ci` and `xc`)?

If `ci` is supposed to work then shouldn't `ca` work too as it is part of the `ci` cast path?

**Comment by Simn** [comment]
To be frank, I have no idea.

Originally, casts were supposed to not be transitive period. This was then softened to deal with some variance situations, but I don't remember the details (I think it was something with `UInt` of all things). There are also exceptions for `@:coreType` types in that regard. And to make it worse, top-down inference might use `from` to influence typing.

I think the only one who ever really had a grasp of this is @frabbit, but he's not using Haxe much these days.

**Comment by vonagam** [comment]
So, a lot of things are depending on transitive direct casts being allowed. Then what is holding back from officially supporting them?

I see one case in [`MyAbstract`](https://github.com/HaxeFoundation/haxe/blob/5fad913a1c843fe413ae3df214ed137c60f764f4/tests/unit/src/unit/MyAbstract.hx#L36-L56): there are abstracts `Kilometer` and `Meter` that both has direct implicit casts with `Float`. It would be definitely not right to allow assignment of one to another through a transitive cast to a float.

For such destructive cases, where a cast loses some crucial information/dimension, we may make them non-transitive - meaning that you can't have more than one such cast in a chain. And one will need to mark abstracts with those casts by some new meta, for example:

```haxe
@:abstractSpace(MyAbstract.length)
abstract Meter(Float) from Float to Float {}

@:abstractSpace(MyAbstract.length)
abstract Kilometer(Float) from Float to Float {}
```

This is just an idea about formalising and making sense of current actual rules.

Are there any other instances with forbidden implicit direct casts which `abstractSpace` or `abstractScope` meta (working name) won't cover?

**Comment by Simn** [comment]
This example tells me that we shouldn't allow such casts by default. If anything, transitive casts should be opt-in via some metadata.

It also helps answering your original question if we modify it to use these types:

```haxe
abstract Meter(Float) from Float to Float {}

abstract Kilometer(Float) {
	@:from static function fromFloat(value:Float):Kilometer {
		return cast value;
	}
}

function main() {
	var distanceFloat = 1.0;
	var distanceMeter:Meter = distanceFloat;
	var distanceKilometer:Kilometer = distanceMeter;
}
```

This `fromFloat` conveys the intent of converting a floating point number that is understood as a kilometer dimension. If I was writing code like that, I'd expect the compiler to admit an actual `Float` value. It's then my responsibility to ensure that this value is actually in the right dimension.

However, I'd certainly _not_ expect it to allow something that is known to be `Meter` to be assignable. If I was being an idiot and had `from Meter` then sure, or if I was less of an idiot and had a `fromMeter` with the appropriate conversion. But otherwise, this should be a compilation error.

My tentative conclusion here is that we should, for a given unification that involves two abstracts, allow _either_ `from` or `to` unification, but not both directions at the same time. I'll note though that this is very much in the narrow mindset of this particular example, so there very well might be type theory that contradicts me.

**Comment by vonagam** [comment]
To clarify last paragraph: are you saying that multiple casts in the same direction are ok, like two consecutive `from`?

**Comment by Simn** [comment]
I still don't think that's a good idea, it opens the possibility of having multiple paths, leading to a big mess.

**Comment by vonagam** [comment]
Have problematic case for implementing said restrictions: [here](https://github.com/HaxeFoundation/haxe/blob/development/tests/unit/src/unit/issues/Issue3494.hx#L29) to swap `T1` and `T2` `E3` converts first to `T1` and then back (now as `T2`), so two casts. Adding whitelist metadata to `E3` allows it, but the question is: is this supposed to work without said meta?

If this is supposed to work as is, then i guess we can make a type constraint cast a free one.
If adding meta to `E3` is ok, then there is no problem at all.

**Comment by Simn** [comment]
The entire `E3` abstract looks like the final boss of all casts...

Let me try to follow the unification here upon dealing with the (implied) `v1 = E3._from2(v2)` call:

Initial type variables:

```
T1
T2
_T1 constrained to E3<T1, T2>
_T2 constrained to E3<T1, T2>
```

Assigning `v2` to the argument `e`:

```
_T1 = String
_T2 = Int
```

Solving constraints:

```
String to E3<T1, T2>
T1 = String via `from T1`
Int to E3<T1, T2>
T2 = Int via `from T2`
```

So at this point we have:

```
T1 = String
T2 = Int
_T1 = String
_T2 = Int
```

This would mean that the return type should be `E3<String, Int>`, but it's actually `E3<Int, String>`. This hinges entirely on the implementation detail that the `from T2` is actually checked before the `from T1`, which suggests that we're missing a `List.rev` somewhere.

So basically, I'm thinking that this code shouldn't compile as-is.

**Comment by Simn** [comment]
I found #7656 via `js.Promise`.

**Comment by Simn** [comment]
I've fixed the order, let's see how many codebases burn now.

As for the actual issue, I don't see why the restriction I mentioned should affect this. To be clear, we have to distinguish two cases here:

1. `A<B> to C to D` casting is transitive and should not be allowed
2. `A<B> to C<D>` casting is fine (given acceptable variance) because it consists of two non-transitive casts: One from `A to C` and one from `B to D`.

We should be dealing with the latter case in this example.
